### PR TITLE
Re-land"[new feature]introduce `ScrollMetricsNotification`"

### PR DESCRIPTION
**Pull request by xu-baolin**
Re-land flutter/flutter#85221

Do not merge before @Piinks LGTMing.

### CODE REVIEW

**Comment by Piinks** [comment]
Currently investigating test failures, I thought they had all passed before we landed the original. I'll find out where the issue is. :)

**Comment by Piinks** [comment]
Update: Still working with customers to fix tests. I was out of town for a bit. :)

**Comment by Piinks** [comment]
Almost there, I think I have it down to one last customer that is using a catch-all NotificationListener. Just working with them to adjust according to the desired behavior.

**Comment by Piinks** [comment]
Scratch that, resolving one test failure reveals more, still digging. I am going to recruit additional help come Monday.

**Comment by Piinks** [comment]
There are a number of tests where `'!renderObject.debugNeedsPaint': is not true.` or `pumpAndSettle timed out`, weren't there failures like this in cocoon originally with this change?

**Comment by xu-baolin** [comment]
> There are a number of tests where `'!renderObject.debugNeedsPaint': is not true.` or `pumpAndSettle timed out`, weren't there failures like this in cocoon originally with this change?

Maybe it has something to do with this,"The new events trigger the remote `Material` widget to re-layout!
https://github.com/flutter/flutter/blob/57a565c70632c9f7001ddf53fe3d1b9ee4d20844/packages/flutter/lib/src/material/material.dart#L379"

**Comment by Piinks** [comment]
Ah I think you are right! I hd not thought to look back in the framework.

File: packages/flutter/test/widgets/scroll_notification_test.dart [hunk 1]
```diff
@@ -7,6 +7,64 @@ import 'package:flutter/material.dart';
...
+          events.add(value);
+          return false;
+        },
+        child: SingleChildScrollView(
+          child: SizedBox(height: height),
+        ),
+      );
+    }
+    await tester.pumpWidget(buildFrame(1200.0));
+    // Initial metrics notification.
```
Comment (xu-baolin) [inline packages/flutter/test/widgets/scroll_notification_test.dart:24]: @Piinks maybe cancel the initial metrics event could resolve the tests, I am not sure.

File: packages/flutter/lib/src/widgets/scroll_position.dart [hunk 2]
```diff
@@ -537,7 +551,14 @@ abstract class ScrollPosition extends ViewportOffset with ScrollMetrics {
       _pendingDimensions = false;
     }
     assert(!_didChangeViewportDimensionOrReceiveCorrection, 'Use correctForNewDimensions() (and return true) to change the scroll offset during applyContentDimensions().');
-    _lastMetrics = copyWith();
+
+    if (_isMetricsChanged()) {
+      // It isn't safe to trigger the ScrollMetricsNotification if we are in
+      // the middle of rendering the frame, the developer is likely to schedule
+      // a new frame(build scheduled during frame is illegal).
+      scheduleMicrotask(didUpdateScrollMetrics);
```
Comment (xu-baolin) [inline packages/flutter/lib/src/widgets/scroll_position.dart:559]:
```suggestion
      if (_lastMetrics != null)
        scheduleMicrotask(didUpdateScrollMetrics);
```

File: packages/flutter/test/widgets/scroll_notification_test.dart [hunk 3]
```diff
@@ -7,6 +7,64 @@ import 'package:flutter/material.dart';
...
+    expect(events.length, 1);
+    ScrollMetricsNotification event = events[0] as ScrollMetricsNotification;
+    expect(event.metrics.extentBefore, 0.0);
+    expect(event.metrics.extentInside, 600.0);
+    expect(event.metrics.extentAfter, 600.0);
```
Comment (xu-baolin) [inline packages/flutter/test/widgets/scroll_notification_test.dart:29]:
```suggestion
    expect(events.length, 0);
```

**Comment by xu-baolin** [comment]
Adjusted the notification sending strategy:
1. Consumers are not notified when `_lastMetrics` is initialized for the first time（from null to non-null）.
2. Merge multiple change events in the same frame, and only notify the last result.

File: packages/flutter/lib/src/widgets/scroll_position.dart [hunk 4]
```diff
@@ -537,7 +553,17 @@ abstract class ScrollPosition extends ViewportOffset with ScrollMetrics {
       _pendingDimensions = false;
     }
     assert(!_didChangeViewportDimensionOrReceiveCorrection, 'Use correctForNewDimensions() (and return true) to change the scroll offset during applyContentDimensions().');
-    _lastMetrics = copyWith();
+
+    if (_isMetricsChanged()) {
+      // It isn't safe to trigger the ScrollMetricsNotification if we are in
+      // the middle of rendering the frame, the developer is likely to schedule
+      // a new frame(build scheduled during frame is illegal).
+      if (_lastMetrics != null && !_haveScheduledUpdateNotification) {
```
Comment (xu-baolin) [inline packages/flutter/lib/src/widgets/scroll_position.dart:561]: Adjusted the notification sending strategy:
1. Consumers are not notified when `_lastMetrics` is initialized for the first time.
2. Merge multiple change events in the same frame, and only notify the last result.

**Comment by Piinks** [comment]
Thanks for the update! I am running this latest version through internal testing. :)

**Comment by Piinks** [comment]
Very close, there are only 9 broken tests with the last patch (f4d2c79), working on debugging them now.

**Comment by Piinks** [comment]
Down to 5, almost there. Thanks for your patience @xu-baolin!

**Comment by Piinks** [comment]
I've been away for family, but I am back now. There are still a few of those customers I need to work with to get tests updated.

File: packages/flutter/lib/src/widgets/scroll_position.dart [hunk 5]
```diff
@@ -942,3 +976,99 @@ abstract class ScrollPosition extends ViewportOffset with ScrollMetrics {
...
+///             ),
+///           ),
+///         ),
+///       ),
+///     );
+///   }
+/// }
+/// ```
+/// {@end-tool}
+class ScrollMetricsNotification extends LayoutChangedNotification with ViewportNotificationMixin {
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/scroll_position.dart:1052]:
```suggestion
class ScrollMetricsNotification extends Notification with ViewportNotificationMixin {
```

I think we should just make this a generic Notification for now. This should allow us to land this change in a non-breaking way. I've tried it against the internal tests and it passed without needing additional customer changes. What do you think @xu-baolin?

Comment (xu-baolin) [reply packages/flutter/lib/src/widgets/scroll_position.dart]: Sounds great, and it makes sense to separate it from `LayoutChangedNotification`.

Comment (Piinks) [reply packages/flutter/lib/src/widgets/scroll_position.dart]: Thank you! Running another round of checks today just to double check again.

**Review by Piinks — APPROVED** [review]
LGTM - triggering additional testing

**Comment by Piinks** [comment]
I've double checked and this is good to go now. Some flaky tests came up when running internal tests, will note in case someone tries to revert. Thanks @xu-baolin!

**Comment by Piinks** [comment]
🎉

**Pull request merged**